pe-parser-library/src/parse.cpp

```diff
@@ -1806,6 +1812,105 @@ bool getRelocations(parsed_pe *p) {
   return true;
 }
 
+bool getDebugDir(parsed_pe *p) {
+  data_directory debugDir;
+  if (p->peHeader.nt.OptionalMagic == NT_OPTIONAL_32_MAGIC) {
+    debugDir = p->peHeader.nt.OptionalHeader.DataDirectory[DIR_DEBUG];
+  } else if (p->peHeader.nt.OptionalMagic == NT_OPTIONAL_64_MAGIC) {
+    debugDir = p->peHeader.nt.OptionalHeader64.DataDirectory[DIR_DEBUG];
+  } else {
+    return false;
+  }
+
+  if (debugDir.Size != 0) {
+    section d;
+    VA vaAddr;
+    if (p->peHeader.nt.OptionalMagic == NT_OPTIONAL_32_MAGIC) {
+      vaAddr =
+          debugDir.VirtualAddress + p->peHeader.nt.OptionalHeader.ImageBase;
+    } else if (p->peHeader.nt.OptionalMagic == NT_OPTIONAL_64_MAGIC) {
+      vaAddr =
+          debugDir.VirtualAddress + p->peHeader.nt.OptionalHeader64.ImageBase;
+    } else {
+      return false;
+    }
+
+    auto numOfDebugEnts = debugDir.Size / sizeof(debug_dir_entry);
+
+    //
+    // this will return the rdata section, where the debug directories are
+    //
+    if (!getSecForVA(p->internal->secs, vaAddr, d)) {
+      return false;
+    }
+
+    //
+    // get debug directory from this section
+    //
+    auto rvaofft = static_cast<std::uint32_t>(vaAddr - d.sectionBase);
+
+    debug_dir_entry emptyEnt;
+    memset(&emptyEnt, 0, sizeof(debug_dir_entry));
+
+    for (int i = 0; i < numOfDebugEnts; i++) {
+      debug_dir_entry curEnt = emptyEnt;
+
+      READ_DWORD(d.sectionData, rvaofft, curEnt, Characteristics);
+      READ_DWORD(d.sectionData, rvaofft, curEnt, TimeStamp);
+      READ_WORD(d.sectionData, rvaofft, curEnt, MajorVersion);
+      READ_WORD(d.sectionData, rvaofft, curEnt, MinorVersion);
+      READ_DWORD(d.sectionData, rvaofft, curEnt, Type);
+      READ_DWORD(d.sectionData, rvaofft, curEnt, SizeOfData);
+      READ_DWORD(d.sectionData, rvaofft, curEnt, AddressOfRawData);
+      READ_DWORD(d.sectionData, rvaofft, curEnt, PointerToRawData);
+
+      // are all the fields in curEnt null? then we break
+      if (curEnt.SizeOfData == 0 && curEnt.AddressOfRawData == 0 &&
+          curEnt.PointerToRawData == 0) {
+        break;
+      }
+
+      //
+      // Get the address of the data
+      //
+      VA rawData;
+      if (p->peHeader.nt.OptionalMagic == NT_OPTIONAL_32_MAGIC) {
+        rawData =
+            curEnt.AddressOfRawData + p->peHeader.nt.OptionalHeader.ImageBase;
+      } else if (p->peHeader.nt.OptionalMagic == NT_OPTIONAL_64_MAGIC) {
+        rawData =
+            curEnt.AddressOfRawData +
+                  p->peHeader.nt.OptionalHeader64.ImageBase;
+      } else {
+        return false;
+      }
+
+      //
+      // Get the section for the data
+      //
+      section dataSec;
+      if (!getSecForVA(p->internal->secs, rawData, dataSec)) {
+        return false;
+      }
+
+      debugent ent;
+
+      auto dataofft = static_cast<std::uint32_t>(rawData - dataSec.sectionBase);
```

Comment:
Same Q about sanity-checking `dataofft` here (especially since we make a buffer from the computed address later).